packages/react-dom/src/events/ReactDOMUpdateBatching.js

```diff
@@ -39,7 +39,6 @@ function finishEventHandler() {
     // If a controlled event was fired, we may need to restore the state of
     // the DOM node back to the controlled value. This is necessary when React
     // bails out of the update without touching the DOM.
-    flushDiscreteUpdatesImpl();
```

Comment:
This one only kicks in for controlled events like onChange. So I don't think your test would ever trigger this path.

I'm surprised this doesn't fail some test and I wonder what else picks up the slack (or if we're missing a test).

This one does behave like flushSync because we need to flush before we call restoreStateIfNeeded. The purpose of that call is that in case something bails out along the path of the rerender (or if it doesn't rerender at all) then we need to reset it to its previous value. However, if it does rerender and updates to some new value, then we don't need to do that.

I suspect that this change now causes the restoring to happen first. So it always restores the value to what it was before and then rerenders and sets it to its new value.

Our tests probably don't cover this subtle difference. Just that it gets reset if it doesn't rerender. However, there are some observable differences. Like the element's value will be different during effects. If I'm right, in a normal controlled component it would always have the old value on the element in an effect instead of the new value.

Additionally, switching the value back and forth can have unintended browser quirks like selection changing or something.

Reply:
Yeah you're right, I got this confused with the other call. I'm too am a bit alarmed that this didn't break anything.

I'll try to write a test.

Reply:
I wonder, could we move the controlled components restoration to the microtask, too?

Reply:
Yea that would probably work.

Reply:
Since this isn't directly related to this PR, I'll create follow up tasks for these.